phpseclib/Net/SshMsgCustom.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace phpseclib3\Net;
+
+/**
+ * @internal
+ */
+abstract class SshMsgCustom
+{
+    const KEXDH_INIT = 30;
+    const KEXDH_REPLY = 31;
```

Comment:
These were originally in `$this->message_numbers`:

https://github.com/phpseclib/phpseclib/blob/cc65aa3b0e9efd982688f921028cbdb158c2905e/phpseclib/Net/SSH2.php#L1211-L1241

Why were they moved out? Like everything else in that array was put in `phpseclib/Net/SshMsg.php` but not these?

Reply:
The spec does not give this 30-49 a specific value, so I thought it should go in extra with the other 30-49 values.

To compensate for this, I had to update lines like this: `$this->updateLogHistory('SSH_MSG_KEXDH_INIT', 'SSH_MSG_KEX_ECDH_INIT');` to be `$this->updateLogHistory('UNKNOWN (30)', 'SSH_MSG_KEX_ECDH_INIT');`

Reply:
That's kinda a bug in the SSH specs. It's _referenced_ in the original SSH2 RFCs (in particular, [RFC4253](https://datatracker.ietf.org/doc/html/rfc4253)) but it's not actually defined anywhere.

I asked about where it was defined back in 2007, when I first started working on phpseclib, at https://groups.google.com/g/comp.security.ssh/c/f5jNVb0fMe4/m/U93doB0Sfn4J, and was told:

> They don't. It accidentally got left out, I think that when it got
(correctly) removed from [SSH-NUMBERS], it was also mistakenly removed
from [SSH-TRANSPORT] as well.
>
> Irritatingly, the erratum for this also seems to have an error, in that
it lists SSH_MSG_KEXDH_REPLY as being 3 when it's obviously 31. Time
for an erratum for the erratum, I suppose.

Reply:
I went ahead and made this change myself!:

https://github.com/phpseclib/phpseclib/commit/b3f14dee377f107ac3bdf7927312148a492f0d85